### ISSUE TITLE
Feature Request: Federate only selected Post Formats

### ISSUE THREAD

**Issue by jsit** [Automattic/wordpress-activitypub#707]
### What

I would like to federate only certain Post Formats -- for instance, `Status` and `Aside`.

### Why

The `Standard` Post Format is often a long blog post that one might not want to federate. Limiting federation to selected Post Formats would allow a user to only federate `Status` posts, for instance.

### How

A section similar to the "Supported post types" in settings: "Supported post formats".

**Comment by mdingemanse** [comment]
I agree with this — it would be nice to have better mechanisms for opting in / out of federation.

**Comment by janboddez** [comment]
>  Might wanna add just one extra line, check that no error occurred? Or maybe it isn't needed, I didn't actually get to test this.

Head another look and if `$transformer` in `class-factory.php` is "falsey" (e.g., `false` or `null`) then the code will fall back to a default depending on the post type. If it is _not_ an instance of `Base` or any of its subclasses, however, then that bit of code will return a `WP_Error`. Except that the rest of the code then goes on and treats it like a transformer instance. Always.

In my view, any code that calls `\Activitypub\Transformer\Factory::get_transformer()` should _probably_ confirm that it really is a valid transformer, and exit/return early if it isn't.

**Comment by jsit** [comment]
Could a hook or filter be added to allow us to do this ourselves? Or is there an existing one that I’m missing?

**Comment by janboddez** [comment]
Finally got the chance to play with this a bit.

Used a must-use (test) plugin like this:
```
<?php

add_filter( 'activitypub_transformer', function ( $transformer, $object, $class ) {

  // By default, `$transformer` equals `null`. *Unless* we return a "truthy"
  // value, the ActivityPub plugin will fall back to a default transformer.
  // Unfortunately, returning anything other than a valid transformer will
  // lead to a fatal PHP error.

  // Assume this worked, though, and we did not want to federate posts in the
  // "Disable ActivityPub" category (or a certain post format, etc.).

  if ( $object instanceof \WP_Post && in_category( 'disable-activitypub', $object ) ) {
    return 'Random string'; // This will cause `\Activitypub\Transformer\Factory::get_transormer()` to return a `\WP_Error`.
                            // Unfortunately, the rest of the code doesn't actually
                            // detect for this, will choke and throw a fatal error.
                            // If there were proper error checks, we could return
                            // random "truthy" values and effectively disable federation
                            // as well as (in theory) content negotiation.
  }

  return $transformer;

}, 10, 3 );
```

Right now, `get_transformer()` is called in exactly 4 places in the code. If we added the right error checks there, we could use this exact hook to effectively disable ActivityPub for certain terms, etc. (Of course, there _might_ be other challenges, like possible race conditions; if so, those would have to be tackled separately.)

**Comment by janboddez** [comment]
> If we added the right error checks there, we could use this exact hook to effectively disable ActivityPub for certain terms, etc.

Like, if any filter callback returned `null` (or if there isn't any callback), that could be interpreted as "use the default." If a custom transformer was returned, of course it should be used instead.

If a callback returned, e.g., `false`, however, or an error, this could be seen as "do not federate."

It wouldn't be too hard to rewrite the code in this manner, and it would be fully backwards compatible.

(In fact, `get_transformer()` is already set up for some kind of "error scenario" even today. But the rest of the code is seemingly unaware.)

**Comment by pfefferle** [comment]
@janboddez good point!

**Comment by vivlim** [comment]
Bump. I installed this on a new blog I set up, but the inability to specify which posts to federate means I have no way to soft-publish posts (e.g. to backfill old ones) without toggling federation globally.

**Comment by janboddez** [comment]
I know it may sound like self promotion, so sorry for that, but would the "local-only category" feature of https://github.com/janboddez/addon-for-activitypub/ work for you? I mean, feel free to test, etc.

Works by, prior to federation (and also as part of the "content negotiation" bit during later lookups), checking against this category (that you can totally set yourself). The code itself is slightly "hacky" in order to make it work with Gutenberg, too. (Which is also why I've been holding off on trying to push it to this plugin itself.) Afraid it's the only way, though.

And it could probably be refactored to work with post formats too, but I haven't got around to it, yet. Just thought maybe the idea of using a category instead of a post format isn't all that weird.

**Comment by jsit** [comment]
Here's my quick-and-dirty solution for only federating "Status" post formats:
```php
add_filter( 'activitypub_send_activity_to_followers', function ( $send, $activity, $actor_id, $outbox_item ) {
  return 'status' === get_post_format( $outbox_item );
}, 10, 4);
```
